state/machine.go

```diff
@@ -905,48 +896,29 @@ func (original *Machine) advanceLifecycle(life Life, force bool, maxWait time.Du
 			var principalUnitNames []string
 			for _, principalUnit := range m.doc.Principals {
 				principalUnitNames = append(principalUnitNames, principalUnit)
-				u, err := m.st.Unit(principalUnit)
-				if err != nil {
-					return nil, errors.Annotatef(err, "reading machine %s principal unit %v", m, m.doc.Principals[0])
-				}
-				app, err := u.Application()
+				canDie, err = m.assessCanDieUnit(principalUnit)
 				if err != nil {
-					return nil, errors.Annotatef(err, "reading machine %s principal unit application %v", m, u.doc.Application)
+					return nil, errors.Trace(err)
 				}
-				if u.Life() == Alive && app.Life() == Alive {
-					canDie = false
+				if !canDie {
 					break
 				}
 			}
 
-			if canDie {
-				containers, err := m.Containers()
-				if err != nil {
-					return nil, errors.Annotatef(err, "reading machine %s containers", m)
-				}
-				canDie = len(containers) == 0
-				containerCheck := txn.Op{
-					C:  containerRefsC,
-					Id: m.doc.DocID,
-					Assert: bson.D{{"$or", []bson.D{
-						{{"children", bson.D{{"$size", 0}}}},
-						{{"children", bson.D{{"$exists", false}}}},
-					}}},
+			if canDie && !dyingAllowContainers {
```

Comment:
Part of the real change to advanceLifecycle allowed machines with container to be marked dying.